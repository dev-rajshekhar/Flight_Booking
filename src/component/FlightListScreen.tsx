import React from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  Modal,
  FlatList,
  ActivityIndicator,
} from 'react-native';
import FlightResponse from '../data/FlightResponse';
import FilterScreen from './FilterScreen';
import FlightItem from './FlightItem';
import {flightResultContainerStyles as styles} from '../style/flight_container';
import colors from '../constants/color';

interface FlightListComponentProps {
  flights: FlightResponse[];
  flightNames: string[];
  selectedFilters: {flightFilter: string; sortOption: string};
  onFilterApply: (selectedFilters: {
    flightFilter: string;
    sortOption: string;
  }) => void;
  onFilterReset: () => void;
  onOpenModal: () => void;
  onCloseModal: () => void;
  filteredFlights: FlightResponse[];
  isModelOpen: boolean;
  onFlightItemClicked: (item: FlightResponse) => void;
  loading: boolean;
  error: any;
}

const FlightListScreen: React.FC<FlightListComponentProps> = ({
  flights,
  selectedFilters,
  onFilterApply,
  onFilterReset,
  onOpenModal,
  onCloseModal,
  flightNames,
  isModelOpen,
  onFlightItemClicked,
  loading,
  error,
}) => {
  const renderFlightItem = ({item}: {item: FlightResponse}) => {
    return (
      <FlightItem
        item={item}
        onFlightItemClicked={onFlightItemClicked}
        isEnabled={true}
      />
    );
  };
  if (loading) {
    return (
      <View style={styles.loadingContainer}>
        <ActivityIndicator size="large" color={colors.negativeButtonColor} />
      </View>
    );
  }
  if (error) {
    return (
      <View style={styles.loadingContainer}>
        <Text>{error}</Text>
      </View>
    );
  }
  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID={'sort-button'}
        style={styles.filterButton}
        onPress={onOpenModal}>
        <Text style={styles.filterButtonText}>Filter & Sort</Text>
      </TouchableOpacity>

      <FlatList
        testID="list"
        data={flights}
        renderItem={renderFlightItem}
        ListEmptyComponent={
          <View style={styles.loadingContainer}>
            <Text>No Records Found</Text>
          </View>
        }
        keyExtractor={item => item.id}
      />
      <Modal visible={isModelOpen} animationType="slide">
        <FilterScreen
          flightName={flightNames}
          selectedFilters={selectedFilters}
          onFilterApply={onFilterApply}
          onFilterReset={onFilterReset}
          onCloseModal={onCloseModal}
        />
      </Modal>
    </View>
  );
};

export default FlightListScreen;
