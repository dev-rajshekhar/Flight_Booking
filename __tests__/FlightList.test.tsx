import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import FlightListScreen from '../src/component/FlightListScreen';

const FlightResponseMock = [
  {
    id: '1',
    fare: 12015,
    displayData: {
      source: {
        airport: [
          {
            cityCode: 'BOM',
            cityName: 'Mumbai',
            terminal: '2',
            airportCode: 'BOM',
            airportName: 'Mumbai',
            countryCode: 'IN',
            countryName: 'India',
          },
        ],
        depTime: '2023-07-18T01:41:13.839116',
      },
      airlines: [
        {
          airlineCode: 'AEC',
          airlineName: 'Vistara',
          flightNumber: '09381',
        },
      ],
      stopInfo: 'Non stop',
      destination: {
        airport: [
          {
            cityCode: 'DEL',
            cityName: 'Delhi',
            terminal: '3',
            airportCode: 'DEL',
            airportName: 'Indira Gandhi Airport',
            countryCode: 'IN',
            countryName: 'India',
          },
        ],
        arrTime: '2023-08-01T01:41:13.839137',
      },
      totalDuration: '3h 54m',
    },
  },
];
jest.mock('../src/component/FlightListScreen', () => ({
  __esModule: true,
  default: jest.fn(() => null),
}));
const mockResponse = {
  flights: FlightResponseMock,
  flightNames: [],
  filteredFlights: FlightResponseMock,
  selectedFilters: {flightFilter: '', sortOption: ''},
  onFilterApply: () => {},
  onFilterReset: () => {},
  onOpenModal: () => {},
  onCloseModal: () => {},
  isModelOpen: false,
};
describe('Should render FlightList Snapshot', () => {
  it('Should match snapshot', () => {
    const component = render(<FlightListScreen {...mockResponse} />).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('Should click on Sort button', () => {
    const sortButton = jest.fn();
    const component = render(<FlightListScreen {...mockResponse} />);
    const buttonComponent = component.getByTestId('sort-button');
    expect(buttonComponent).toBeTruthy();
    fireEvent.press(buttonComponent);
    expect(sortButton).toHaveBeenCalledTimes(1);
  });
});
