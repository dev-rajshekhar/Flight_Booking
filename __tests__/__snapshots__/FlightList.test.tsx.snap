// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render FlightList Snapshot Should match snapshot 1`] = `null`;
